File: src/dotnet/Core/Application/Services/TechnologyMatcher.cs
using Rag.Candidates.Core.Domain.Configuration;
using System.Text.RegularExpressions;

namespace Rag.Candidates.Core.Application.Services;

public sealed class TechnologyMatcher
{
    private readonly QueryParsingConfig _config;

    public TechnologyMatcher(QueryParsingConfig config)
    {
        _config = config;
    }

    public string[] ExtractTechnologies(string queryText)
    {
        var technologies = new HashSet<string>();
        var queryLower = queryText.ToLowerInvariant();

        foreach (var (token, normalized) in _config.TechnologySynonyms)
        {
            var pattern = $@"\b{Regex.Escape(token.ToLowerInvariant())}\b";
            if (Regex.IsMatch(queryLower, pattern))
            {
                technologies.Add(normalized);
            }
        }

        return technologies.OrderBy(t => t).ToArray();
    }
}

File: src/python/core/application/services/technology_matcher.py
from typing import List
import re
from ...domain.configuration.query_parsing_config import QueryParsingConfig


class TechnologyMatcher:
    def __init__(self, config: QueryParsingConfig):
        self._config = config
    
    def extract_technologies(self, query_text: str) -> List[str]:
        technologies = []
        query_lower = query_text.lower()
        
        matched_tokens = set()
        
        for token, normalized in self._config.technology_synonyms.items():
            pattern = r'\b' + re.escape(token.lower()) + r'\b'
            if re.search(pattern, query_lower):
                matched_tokens.add(normalized)
        
        return sorted(list(matched_tokens))
